compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/AnnotationDesugar.java

```diff
@@ -191,6 +199,84 @@ private void defineClassAnnotations(BLangPackage pkgNode, SymbolEnv env, BLangFu
+        if (annSymbol instanceof BAnnotationSymbol) {
+            annoAttachment.annotationSymbol = (BAnnotationSymbol) annSymbol;
+        }
+
+        BLangIdentifier identifierNode = (BLangIdentifier) TreeBuilder.createIdentifierNode();
+        annoAttachment.annotationName = identifierNode;
+        annoAttachment.annotationName.value = SERVICE_INTROSPECTION_INFO_ANN;
+        annoAttachment.pos = position;
+        annoAttachment.annotationName.pos = position;
+
+        // create record literal related to annotation node
+        BLangRecordLiteral literalNode = (BLangRecordLiteral) TreeBuilder.createRecordLiteralNode();
+        annoAttachment.expr = literalNode;
+        BLangIdentifier pkgAlias = (BLangIdentifier) TreeBuilder.createIdentifierNode();
+        pkgAlias.setValue(BUILTIN_PKG_KEY);
+        annoAttachment.pkgAlias = pkgAlias;
+        annoAttachment.attachPoints.add(AttachPoint.Point.SERVICE);
+        literalNode.pos = position;
+        BSymbol annTypeSymbol = symResolver.lookupSymbolInMainSpace(
+                pkgEnv, names.fromString(SERVICE_INTROSPECTION_INFO_REC));
+        BStructureTypeSymbol bStructSymbol = null;
+        if (annTypeSymbol instanceof BStructureTypeSymbol) {
+            bStructSymbol = (BStructureTypeSymbol) annTypeSymbol;
+            literalNode.setBType(bStructSymbol.type);
+        }
+
+        //Add Root Descriptor
+        BLangRecordLiteral.BLangRecordKeyValueField descriptorKeyValue = (BLangRecordLiteral.BLangRecordKeyValueField)
+                TreeBuilder.createRecordKeyValue();
+        literalNode.fields.add(descriptorKeyValue);
+
+        // create annotation field key for `name`
+        BLangLiteral keyLiteral = (BLangLiteral) TreeBuilder.createLiteralExpression();
+        keyLiteral.value = SERVICE_NAME;
+        keyLiteral.setBType(symTable.stringType);
+
+        // create annotation field value for `name`
+        BLangLiteral valueLiteral = (BLangLiteral) TreeBuilder.createLiteralExpression();
+        valueLiteral.setBType(symTable.stringType);
+        valueLiteral.value = generateServiceHashCode(serviceClass);
+        valueLiteral.pos = position;
+
+        descriptorKeyValue.key = new BLangRecordLiteral.BLangRecordKey(keyLiteral);
+        BSymbol fieldSymbol = symResolver.resolveStructField(position, pkgEnv,
+                names.fromString(SERVICE_NAME), bStructSymbol);
+
+        if (fieldSymbol instanceof BVarSymbol) {
+            descriptorKeyValue.key.fieldSymbol = (BVarSymbol) fieldSymbol;
+        }
+        if (valueLiteral != null) {
+            descriptorKeyValue.valueExpr = valueLiteral;
+        }
+
+        // add generated annotation to the service definition
+        serviceClass.addAnnotationAttachment(annoAttachment);
+    }
+
+    private String generateServiceHashCode(BLangClassDefinition serviceClass) {
+        String serviceName = "service$" + serviceClass.name.getValue();
```

Comment:
Is this name you generate and the name you use in the hash code impl of `BallerinaServiceDeclarationSymbol` the same? 🤔

Reply:
Yes, `serviceClass.name.getValue()` just have the `$anonType$_` + `<auto-incremented-value>` value. So, we prepend `service$` part.